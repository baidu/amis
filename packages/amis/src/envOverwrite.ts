/**
 * @file 用于在移动端或不同语言环境下使用不同配置
 */

import {SchemaNode, Schema} from './types';
import {RendererProps, RendererConfig, addSchemaFilter} from './factory';
import {findObjectsWithKey} from './utils/helper';
const isMobile = (window as any).matchMedia?.('(max-width: 768px)').matches
  ? true
  : false;

// 这里不能用 addSchemaFilter 是因为还需要更深层的替换，比如 select 里的 options
export const envOverwrite = (schema: any, locale?: string) => {
  if (schema.mobile && isMobile) {
    Object.assign(schema, schema.mobile);
    delete schema.mobile;
  }

  if (locale) {
    let schemaNodes = findObjectsWithKey(schema, locale);
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode[locale]);
      delete schemaNode[locale];
    }
  }

  if (isMobile) {
    let schemaNodes = findObjectsWithKey(schema, 'mobile');
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode['mobile']);
      delete schemaNode['mobile'];
    }
  }
};
