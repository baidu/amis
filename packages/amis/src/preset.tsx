import {
  addRootWrapper,
  extendDefaultEnv,
  LazyComponent,
  render,
  themeable,
  ThemeProps
} from 'amis-core';

import {alert, confirm, setRenderSchemaFn, toast, ImageGallery} from 'amis-ui';
import React from 'react';

extendDefaultEnv({
  alert,
  confirm,
  notify: (type: keyof typeof toast, msg: string, conf: any) =>
    toast[type] ? toast[type](msg, conf) : console.warn('[Notify]', type, msg)
});

setRenderSchemaFn((controls, value, callback, scopeRef, theme) => {
  return render(
    {
      name: 'form',
      type: 'form',
      wrapWithPanel: false,
      mode: 'horizontal',
      controls,
      messages: {
        validateFailed: ''
      }
    },
    {
      data: value,
      onFinished: callback,
      scopeRef,
      theme
    },
    {
      session: 'prompt'
    }
  );
});

addRootWrapper((props: any) => {
  const {env, children} = props;
  return (
    <ImageGallery modalContainer={env.getModalContainer}>
      {children}
    </ImageGallery>
  );
});

const SimpleSpinner = themeable(
  (
    props: {
      className?: string;
      spinnerClassName?: string;
    } & ThemeProps
  ) => {
    const cx = props.classnames;

    return (
      <div
        data-testid="spinner"
        className={cx(`Spinner`, 'in', props.className)}
      >
        <div
          className={cx(
            `Spinner-icon`,
            'Spinner-icon--default',
            props.spinnerClassName
          )}
        ></div>
      </div>
    );
  }
);

LazyComponent.defaultProps.placeholder = <SimpleSpinner />;
