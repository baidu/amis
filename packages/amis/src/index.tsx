/** @license amis v@version
 *
 * Copyright Baidu
 *
 * This source code is licensed under the Apache license found in the
 * LICENSE file in the root directory of this source tree.
 */
export * from 'amis-core';
export * from 'amis-ui';

import './preset';

// 注册渲染器
import './renderers/Action';
import './renderers/Alert';
import './renderers/App';
import './renderers/Avatar';
import './renderers/Remark';
import './renderers/ButtonGroup';
import './renderers/Form/ButtonToolbar';
import './renderers/Breadcrumb';
import './renderers/DropDownButton';
import './renderers/Calendar';
import './renderers/Collapse';
import './renderers/CollapseGroup';
import './renderers/Color';
import './renderers/CRUD';
import './renderers/CRUD2';
import './renderers/Pagination';
import './renderers/Cards';
import './renderers/Card';
import './renderers/Card2';
import './renderers/Custom';
import './renderers/Date';
import './renderers/Dialog';
import './renderers/Divider';
import './renderers/Each';
import './renderers/Flex';
import './renderers/Form/Control';
import './renderers/Form/Hidden';
import './renderers/Form/InputText';
import './renderers/Form/InputTag';
import './renderers/Form/InputNumber';
import './renderers/Form/Textarea';
import './renderers/Form/Checkboxes';
import './renderers/Form/Checkbox';
import './renderers/Form/InputCity';
import './renderers/Form/ChartRadios';
import './renderers/Form/InputRating';
import './renderers/Form/Switch';
import './renderers/Form/Radios';
import './renderers/Form/Radio';
import './renderers/Form/JSONSchema';
import './renderers/Form/JSONSchemaEditor';
import './renderers/Form/ListSelect';
import './renderers/Form/LocationPicker';
import './renderers/Form/Select';
import './renderers/Form/Static';
import './renderers/Form/InputDate';
import './renderers/Form/InputDateRange';
import './renderers/Form/InputFormula';
import './renderers/Form/InputRepeat';
import './renderers/Form/InputTree';
import './renderers/Form/TreeSelect';
import './renderers/Form/InputImage';
import './renderers/Form/InputFile';
import './renderers/Form/UUID';
import './renderers/Form/MatrixCheckboxes';
import './renderers/Form/InputMonthRange';
import './renderers/Form/InputQuarterRange';
import './renderers/Form/InputYearRange';
import './renderers/Form/InputRange';
import './renderers/Form/InputArray';
import './renderers/Form/Combo';
import './renderers/Form/ConditionBuilder';
import './renderers/Form/InputSubForm';
import './renderers/Form/InputExcel';
import './renderers/Form/InputRichText';
import './renderers/Form/Editor';
import './renderers/Form/DiffEditor';
import './renderers/Form/InputColor';
import './renderers/Form/ChainedSelect';
import './renderers/Form/NestedSelect';
import './renderers/Form/Transfer';
import './renderers/Form/TransferPicker';
import './renderers/Form/InputTable';
import './renderers/Form/Picker';
import './renderers/Form/IconPicker';
import './renderers/Form/IconSelect';
import './renderers/Form/Formula';
import './renderers/Form/FieldSet';
import './renderers/Form/TabsTransfer';
import './renderers/Form/TabsTransferPicker';
import './renderers/Form/Group';
import './renderers/Form/InputGroup';
import './renderers/Form/UserSelect';
import './renderers/Grid';
import './renderers/Grid2D';
import './renderers/HBox';
import './renderers/VBox';
import './renderers/Image';
import './renderers/Images';
import './renderers/List';
import './renderers/Log';
import './renderers/Operation';
import './renderers/Page';
import './renderers/PaginationWrapper';
import './renderers/Panel';
import './renderers/Plain';
import './renderers/Property';
import './renderers/Portlet';
import './renderers/Spinner';
import './renderers/Table/index';
import './renderers/Tabs';
import './renderers/Tpl';
import './renderers/Mapping';
import './renderers/Progress';
import './renderers/Status';
import './renderers/Json';
import './renderers/Link';
import './renderers/Wizard';
import './renderers/Chart';
import './renderers/Container';
import './renderers/SwitchContainer';
import './renderers/SearchBox';
import './renderers/Service';
import './renderers/SparkLine';
import './renderers/Video';
import './renderers/Audio';
import './renderers/Nav';
import './renderers/Number';
import './renderers/Tasks';
import './renderers/Drawer';
import './renderers/Wrapper';
import './renderers/IFrame';
import './renderers/BarCode';
import './renderers/QRCode';
import './renderers/Icon';
import './renderers/Carousel';
import './renderers/AnchorNav';
import './renderers/Steps';
import './renderers/Timeline';
import './renderers/Markdown';
import './renderers/TableView';
import './renderers/Code';
import './renderers/WebComponent';
import './renderers/GridNav';
import './renderers/TooltipWrapper';
import './renderers/Tag';
import './renderers/Table2/index';
import './renderers/Words';
import './renderers/Password';
import './renderers/DateRange';
import './renderers/MultilineText';
import './renderers/OfficeViewer';
import './renderers/AMIS';

import './compat';
import './schemaExtend';
import type {
  BaseSchema,
  FormSchema,
  SchemaApi,
  SchemaCollection,
  SchemaExpression,
  SchemaObject
} from './Schema';
import type {TableViewSchema, TrObject} from './renderers/TableView';
import type {ActionSchema, ButtonSchema} from './renderers/Action';
import type {CRUDCommonSchema} from './renderers/CRUD';
import type {CRUD2Schema} from './renderers/CRUD2';
import type {TabsSchema} from './renderers/Tabs';
import {availableLanguages as EditorAvailableLanguages} from './renderers/Form/Editor';
import type {Action} from './types';
export * from './renderers/Form/IconPickerIcons';
export * from './renderers/Form/IconSelectStore';

export {
  BaseSchema,
  SchemaCollection,
  FormSchema,
  SchemaApi,
  SchemaObject,
  TableViewSchema,
  TrObject,
  ActionSchema,
  CRUDCommonSchema,
  ButtonSchema,
  CRUD2Schema,
  TabsSchema,
  SchemaExpression,
  Action,
  EditorAvailableLanguages
};
