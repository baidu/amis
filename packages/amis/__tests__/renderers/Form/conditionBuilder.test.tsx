/**
 * 组件名称：Checkboxes 复选框
 * 单测内容：
 * 1. 基础使用
 * 2. 添加条件
 * 3. 拖拽排序
 * 4. 数字类型
 * 5. select 类型、select 配置 select 与 searchable
 * 6. 自定义字段
 * 7. 远程拉取字段 source
 * 8. selectMode
 * 9. builderMode & showANDOR & showNot
 * 10. 非内嵌模式
 */

import React from 'react';
import {fireEvent, render, screen, waitFor} from '@testing-library/react';
import '../../../src';
import {render as amisRender} from '../../../src';
import {makeEnv, replaceReactAriaIds, wait} from '../../helper';

const testSchema = {
  type: 'page',
  body: {
    type: 'form',
    submitText: 'Submit',
    body: [
      {
        type: 'condition-builder',
        label: '条件组件',
        changeImmediately: true,
        name: 'conditions',
        description:
          '适合让用户自己拼查询条件，然后后端根据数据生成 query where',
        fields: [
          {
            label: '文本',
            type: 'text',
            name: 'text'
          },
          {
            label: '数字',
            type: 'number',
            name: 'number'
          },
          {
            label: '布尔',
            type: 'boolean',
            name: 'boolean'
          },
          {
            label: '选项',
            type: 'select',
            name: 'select',
            options: [
              {
                label: 'A',
                value: 'a'
              },
              {
                label: 'B',
                value: 'b'
              },
              {
                label: 'C',
                value: 'c'
              },
              {
                label: 'D',
                value: 'd'
              },
              {
                label: 'E',
                value: 'e'
              }
            ]
          },
          {
            label: '动态选项',
            type: 'select',
            name: 'select2',
            searchable: true,
            placeholder: '这个是下拉框',
            source: '/api/mock2/form/getOptions?waitSeconds=1'
          },
          {
            label: '日期',
            children: [
              {
                // 方便获取
                label: '日期测试',
                type: 'date',
                name: 'date'
              },
              {
                label: '时间',
                type: 'time',
                name: 'time'
              },
              {
                label: '日期时间',
                type: 'datetime',
                name: 'datetime',
                format: 'YYYY-MM-DD hh:mm:ss'
              }
            ]
          }
        ]
      }
    ]
  }
};

test('Renderer:condition-builder', () => {
  const {container} = render(amisRender(testSchema, {}, makeEnv({})));

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
});

test('Renderer:condition-builder add', async () => {
  const onSubmit = jest.fn();
  const {container, findByText, findByPlaceholderText} = render(
    amisRender(testSchema, {onSubmit}, makeEnv({}))
  );

  const andCondition = await findByText('且');

  fireEvent.mouseOver(andCondition);

  const addCondition = await findByText('添加条件');

  fireEvent.click(addCondition);

  const inputText = await findByText('请选择字段');

  fireEvent.click(inputText);

  const textType = await findByText('文本');

  fireEvent.click(textType);

  const textOpType = await findByText('请选择操作');

  fireEvent.click(textOpType);

  const qualOpType = await findByText('等于');

  fireEvent.click(qualOpType);

  const textRightInput = await findByPlaceholderText('请输入文本');

  fireEvent.change(textRightInput, {target: {value: 'amis'}});

  fireEvent.click(await findByText('Submit'));
  await wait(200);

  /** Form的debug区域升级成json组件了，无法直接通过innerHtml获取form data */
  const formData = onSubmit.mock.calls[0][0];
  expect(onSubmit).toHaveBeenCalled();
  expect(formData).toMatchObject({
    conditions: {
      conjunction: 'and',
      children: [
        {
          left: {
            type: 'field',
            field: 'text'
          },
          op: 'equal',
          right: 'amis'
        }
      ]
    }
  });
});

test('Renderer:condition-builder drag order', async () => {
  const {container, findByText} = render(
    amisRender(
      {
        type: 'page',
        body: {
          type: 'form',
          body: [
            {
              type: 'condition-builder',
              label: '条件组件',
              name: 'conditions',
              description:
                '适合让用户自己拼查询条件，然后后端根据数据生成 query where',
              fields: [
                {
                  label: 'AText',
                  type: 'text',
                  name: 'a'
                },
                {
                  label: 'BText',
                  type: 'text',
                  name: 'b'
                }
              ],
              value: {
                id: 'ab8c8eaea7e4',
                conjunction: 'and',
                children: [
                  {
                    id: '87cbc666c5ce',
                    left: {
                      type: 'field',
                      field: 'a'
                    }
                  },
                  {
                    id: '395df0331a46',
                    left: {
                      type: 'field',
                      field: 'b'
                    }
                  }
                ]
              }
            }
          ]
        }
      },
      {},
      makeEnv({})
    )
  );

  await findByText('BText');

  const dragbar = container.querySelectorAll('.cxd-CBGroupOrItem-dragbar');

  // TODO: jsdom 目前还不支持 drag，用不了
  // fireEvent.dragStart(dragbar[1]);
  // fireEvent.dragEnter(dragbar[0]);
  // fireEvent.drop(dragbar[0]);
  // fireEvent.dragLeave(dragbar[0]);
  // fireEvent.dragEnd(dragbar[1]);
});

test('Renderer:condition-builder with number type', async () => {
  const onSubmit = jest.fn();
  const {container, findByText, findByPlaceholderText} = render(
    amisRender(testSchema, {onSubmit}, makeEnv({}))
  );

  fireEvent.click(await findByText('添加条件'));
  fireEvent.click(await findByText('请选择字段'));
  fireEvent.click(await findByText('数字'));
  fireEvent.click(await findByText('请选择操作'));
  fireEvent.click(await findByText('等于'));

  fireEvent.change(await findByPlaceholderText('请输入数字'), {
    target: {value: 81192}
  });

  const SubmitAndCheck = async (calledTime: number, result: any) => {
    fireEvent.click(await findByText('Submit'));
    await wait(200);

    expect(onSubmit).toBeCalledTimes(calledTime);
    expect(onSubmit.mock.calls[calledTime - 1][0]!.conditions).toMatchObject(
      result
    );
  };
  await SubmitAndCheck(1, {
    conjunction: 'and',
    children: [
      {
        left: {
          type: 'field',
          field: 'number'
        },
        op: 'equal',
        right: 81192
      }
    ]
  });

  fireEvent.click(await findByText('等于'));
  fireEvent.click(await findByText('小于'));
  fireEvent.change(await findByPlaceholderText('请输入数字'), {
    target: {value: 1233}
  });

  await SubmitAndCheck(2, {
    conjunction: 'and',
    children: [
      {
        left: {
          type: 'field',
          field: 'number'
        },
        op: 'less',
        right: 1233
      }
    ]
  });

  fireEvent.click(await findByText('小于'));
  fireEvent.click(await findByText('大于或等于'));
  fireEvent.change(await findByPlaceholderText('请输入数字'), {
    target: {value: 1688}
  });

  await SubmitAndCheck(3, {
    conjunction: 'and',
    children: [
      {
        left: {
          type: 'field',
          field: 'number'
        },
        op: 'greater_or_equal',
        right: 1688
      }
    ]
  });

  fireEvent.click(await findByText('大于或等于'));
  fireEvent.click(await findByText('属于范围'));
  const inputs = container.querySelectorAll('.cxd-Number input')!;
  fireEvent.change(inputs[0], {
    target: {value: 11}
  });
  fireEvent.change(inputs[1], {
    target: {value: 22}
  });

  await SubmitAndCheck(4, {
    conjunction: 'and',
    children: [
      {
        left: {
          type: 'field',
          field: 'number'
        },
        op: 'between',
        right: [11, 22]
      }
    ]
  });

  fireEvent.click(await findByText('属于范围'));
  fireEvent.click(await findByText('不为空'));

  await SubmitAndCheck(5, {
    conjunction: 'and',
    children: [
      {
        left: {
          type: 'field',
          field: 'number'
        },
        op: 'is_not_empty',
        right: [11, 22]
      }
    ]
  });
});

test('Renderer:condition-builder with select type & source & searchable', async () => {
  const onSubmit = jest.fn();
  const fetcher = jest.fn().mockImplementationOnce(() =>
    Promise.resolve({
      data: {
        status: 0,
        msg: 'ok',
        data: {
          options: [
            {label: 'Option A', value: 'a'},
            {label: 'Option B', value: 'b'},
            {label: 'Option C', value: 'c'},
            {label: 'Option D', value: 'd'},
            {label: 'Option E', value: 'e'}
          ]
        }
      }
    })
  );
  const {container, findByText, findByPlaceholderText, findByTitle} = render(
    amisRender(
      testSchema,
      {onSubmit},
      makeEnv({
        fetcher,
        session: 'in-condition-builder-1'
      })
    )
  );

  fireEvent.click(await findByText('添加条件'));
  fireEvent.click(await findByText('请选择字段'));
  fireEvent.click(await findByText('动态选项'));
  fireEvent.click(await findByText('请选择操作'));
  fireEvent.click(await findByText('等于'));

  await wait(200);
  expect(fetcher).toHaveBeenCalled();

  fireEvent.click(await findByText('这个是下拉框'));

  expect(
    container.querySelectorAll('.cxd-CBValue .cxd-Select-option')!.length
  ).toBe(5);
  expect(await findByText('Option A')).toBeInTheDocument();

  fireEvent.change(await findByPlaceholderText('搜索'), {
    target: {value: 'C'}
  });

  expect(
    container.querySelectorAll('.cxd-CBValue .cxd-Select-option')!.length
  ).toBe(1);
  // expect(container).toMatchSnapshot();
  expect(await findByTitle('Option C')).toBeInTheDocument();
});

test('Renderer:condition-builder with custom field', async () => {
  const onSubmit = jest.fn();
  const {container, findByText, findByPlaceholderText, findByTitle} = render(
    amisRender(
      {
        type: 'form',
        submitText: 'Submit',
        body: [
          {
            type: 'condition-builder',
            label: '条件组件',
            name: 'conditions',
            description:
              '适合让用户自己拼查询条件，然后后端根据数据生成 query where',
            fields: [
              {
                label: '自定义',
                type: 'custom',
                name: 'a',
                value: {
                  type: 'input-color'
                },
                operators: [
                  'equal',
                  {
                    label: '等于（自定义）',
                    value: 'custom_equal',
                    values: [
                      {
                        type: 'input-color',
                        name: 'color1'
                      },
                      {
                        type: 'tpl',
                        tpl: '~'
                      },
                      {
                        type: 'input-color',
                        name: 'color2'
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      },
      {onSubmit},
      makeEnv({})
    )
  );

  fireEvent.click(await findByText('添加条件'));
  fireEvent.click(await findByText('请选择字段'));
  fireEvent.click(await findByText('自定义'));
  fireEvent.click(await findByText('请选择操作'));
  fireEvent.click(await findByText('等于（自定义）'));

  await wait(400);
  const colorInputs = container.querySelectorAll(
    '.cxd-CBValue .cxd-ColorPicker-input'
  )!;
  expect(colorInputs.length).toBe(2);

  fireEvent.change(colorInputs[0], {
    target: {value: '#101'}
  });
  await wait(200);
  fireEvent.change(colorInputs[1], {
    target: {value: '#f0f'}
  });

  await wait(200);
  fireEvent.click(await findByText('Submit'));
  await wait(200);

  expect(onSubmit).toBeCalledTimes(1);
  expect(onSubmit.mock.calls[0][0]!).toMatchObject({
    conditions: {
      conjunction: 'and',
      children: [
        {
          left: {
            type: 'field',
            field: 'a'
          },
          op: 'custom_equal',
          right: {
            color1: '#101',
            color2: '#f0f'
          }
        }
      ]
    }
  });
  // expect(container).toMatchSnapshot();
});

test('Renderer:condition-builder with source fields', async () => {
  const onSubmit = jest.fn();
  const fetcher = jest.fn().mockImplementationOnce(() =>
    Promise.resolve({
      data: {
        status: 0,
        msg: 'ok',
        data: {
          fields: [
            {
              label: '布尔',
              name: 'boolean',
              type: 'boolean'
            },
            {
              children: [
                {label: '日期', type: 'date', name: 'date'},
                {label: '时间', type: 'time', name: 'time'}
              ],
              label: '日期'
            }
          ]
        }
      }
    })
  );
  const {container, findByText, findByPlaceholderText, findByTitle} = render(
    amisRender(
      {
        type: 'form',
        submitText: 'Submit',
        body: [
          {
            type: 'condition-builder',
            label: '条件组件',
            name: 'conditions',
            source: '/api/condition-fields/custom'
          }
        ]
      },
      {onSubmit},
      makeEnv({
        fetcher,
        session: 'in-condition-builder-2'
      })
    )
  );

  await wait(200);
  expect(fetcher).toHaveBeenCalled();

  fireEvent.click(await findByText('添加条件'));
  fireEvent.click(await findByText('请选择字段'));
  fireEvent.click(await findByText('布尔'));
  fireEvent.click(await findByText('请选择操作'));
  fireEvent.click(await findByText('等于'));
  fireEvent.click(container.querySelector('.cxd-Switch')!);

  await wait(200);
  fireEvent.click(await findByText('Submit'));
  await wait(200);

  expect(onSubmit).toBeCalledTimes(1);
  expect(onSubmit.mock.calls[0][0]!).toMatchObject({
    conditions: {
      conjunction: 'and',
      children: [
        {
          left: {
            type: 'field',
            field: 'boolean'
          },
          op: 'equal',
          right: true
        }
      ]
    }
  });
});

test('Renderer:condition-builder with selectMode', async () => {
  const {container, findByText, findByPlaceholderText, findByTitle} = render(
    amisRender(
      {
        type: 'form',
        submitText: 'Submit',
        body: [
          {
            type: 'condition-builder',
            label: '条件组件',
            name: 'conditions',
            selectMode: 'tree',
            fields: [
              {
                label: '文本',
                type: 'text',
                name: 'text'
              },
              {
                label: '数字',
                type: 'number',
                name: 'number'
              },
              {
                label: '布尔',
                type: 'boolean',
                name: 'boolean'
              },
              {
                label: '树形结构',
                type: 'tree',
                name: 'tree',
                children: [
                  {
                    label: 'Folder A',
                    value: 1,
                    children: [
                      {
                        label: 'file A',
                        value: 2
                      },
                      {
                        label: 'Folder B',
                        value: 3,
                        children: [
                          {
                            label: 'file b1',
                            value: 3.1
                          },
                          {
                            label: 'file b2',
                            value: 3.2
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      },
      {},
      makeEnv({})
    )
  );

  fireEvent.click(await findByText('添加条件'));
  fireEvent.click(await findByText('请选择字段'));

  expect(container.querySelector('.cxd-TreeSelection')).toBeInTheDocument();
  // expect(container).toMatchSnapshot();
});

test('Renderer:condition-builder with builderMode & showANDOR & showNot', async () => {
  const onSubmit = jest.fn();
  const {container, findByText, findByPlaceholderText, findByTitle} = render(
    amisRender(
      {
        type: 'form',
        submitText: 'Submit',
        body: [
          {
            type: 'condition-builder',
            label: '条件组件',
            name: 'conditions',
            builderMode: 'simple',
            showANDOR: true,
            showNot: true,
            fields: testSchema.body.body[0].fields
          }
        ]
      },
      {onSubmit},
      makeEnv({})
    )
  );

  fireEvent.click(await findByText('添加条件'));

  expect(
    container.querySelector('.cxd-CBGroupOrItem-simple')
  ).toBeInTheDocument();

  fireEvent.click(await findByText('请选择字段'));
  fireEvent.click(await findByText('布尔'));
  fireEvent.click(await findByText('请选择操作'));
  fireEvent.click(await findByText('等于'));
  fireEvent.click(container.querySelector('.cxd-Switch')!);

  await wait(200);
  fireEvent.click(await findByText('Submit'));
  await wait(200);

  expect(onSubmit).toBeCalledTimes(1);
  expect(onSubmit.mock.calls[0][0]!).toMatchObject({
    conditions: {
      conjunction: 'and',
      children: [
        {
          left: {
            type: 'field',
            field: 'boolean'
          },
          op: 'equal',
          right: true
        }
      ]
    }
  });

  fireEvent.click(await findByText('且'));
  fireEvent.click(await findByText('或'));

  await wait(200);
  fireEvent.click(await findByText('Submit'));
  await wait(200);

  expect(onSubmit).toBeCalledTimes(2);
  expect(onSubmit.mock.calls[1][0]!).toMatchObject({
    conditions: {
      conjunction: 'or',
      children: [
        {
          left: {
            type: 'field',
            field: 'boolean'
          },
          op: 'equal',
          right: true
        }
      ]
    }
  });

  fireEvent.click(await findByText('非'));

  await wait(200);
  fireEvent.click(await findByText('Submit'));
  await wait(200);

  expect(onSubmit).toBeCalledTimes(3);
  expect(onSubmit.mock.calls[2][0]!).toMatchObject({
    conditions: {
      conjunction: 'or',
      not: true,
      children: [
        {
          left: {
            type: 'field',
            field: 'boolean'
          },
          op: 'equal',
          right: true
        }
      ]
    }
  });
});

test('Renderer:condition-builder with not embed', async () => {
  const onSubmit = jest.fn();
  const {container, findByText, baseElement, findByTitle} = render(
    amisRender(
      {
        type: 'form',
        submitText: 'Submit',
        body: [
          {
            type: 'condition-builder',
            label: '条件组件',
            embed: false,
            title: '条件组合设置',
            placeholder: '自定义占位',
            name: 'conditions',
            fields: testSchema.body.body[0].fields
          }
        ]
      },
      {onSubmit},
      makeEnv({})
    )
  );

  fireEvent.click(await findByText('自定义占位'));
  await wait(300);

  fireEvent.click(await findByText('添加条件'));
  fireEvent.click(await findByText('请选择字段'));
  fireEvent.click(await findByText('日期测试'));
  fireEvent.click(await findByText('请选择操作'));
  fireEvent.click(await findByText('不属于范围'));

  expect(
    baseElement.querySelector('.cxd-Modal .cxd-CBGroup')
  ).toBeInTheDocument();
});
