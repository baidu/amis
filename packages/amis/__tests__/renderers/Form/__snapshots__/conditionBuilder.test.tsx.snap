// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer:condition-builder 1`] = `
<div>
  <div
    class="cxd-Page"
  >
    <div
      class="cxd-Page-content"
    >
      <div
        class="cxd-Page-main"
      >
        <div
          class="cxd-Page-body"
          role="page-body"
        >
          <div
            class="cxd-Panel cxd-Panel--default cxd-Panel--form"
          >
            <div
              class="cxd-Panel-heading"
            >
              <h3
                class="cxd-Panel-title"
              >
                <span
                  class="cxd-TplField"
                >
                  <span>
                    表单
                  </span>
                </span>
              </h3>
            </div>
            <div
              class="cxd-Panel-body"
            >
              <form
                class="cxd-Form cxd-Form--normal"
                novalidate=""
              >
                <input
                  style="display: none;"
                  type="submit"
                />
                <div
                  class="cxd-Form-item cxd-Form-item--normal"
                  data-role="form-item"
                >
                  <label
                    class="cxd-Form-label"
                  >
                    <span>
                      <span
                        class="cxd-TplField"
                      >
                        <span>
                          条件组件
                        </span>
                      </span>
                    </span>
                  </label>
                  <div
                    class="cxd-ConditionBuilderControl cxd-Form-control"
                  >
                    <div
                      class="cxd-CBGroup"
                    >
                      <div
                        class="cxd-CBGroup-toolbarCondition"
                        draggable="true"
                      >
                        <div
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          class="cxd-Select"
                          role="combobox"
                          tabindex="0"
                        >
                          <div
                            class="cxd-Select-valueWrap"
                          >
                            <div
                              class="cxd-Select-value"
                            >
                              且
                            </div>
                          </div>
                          <span
                            class="cxd-Select-arrow"
                          >
                            <icon-mock
                              classname="icon icon-right-arrow-bold"
                              icon="right-arrow-bold"
                            />
                          </span>
                        </div>
                      </div>
                      <div
                        class="cxd-CBGroup-body-wrapper"
                      >
                        <div
                          class="cxd-CBGroup-body"
                        >
                          <div
                            class="cxd-CBGroup-placeholder "
                          >
                            空
                          </div>
                        </div>
                        <div
                          class="cxd-CBGroup-toolbar"
                          draggable="true"
                        >
                          <div
                            class="cxd-CBGroup-toolbarConditionAdd"
                          >
                            <div
                              class="cxd-ButtonGroup"
                            >
                              <button
                                class="cxd-Button cxd-Button--link cxd-Button--size-xs"
                                type="button"
                              >
                                添加条件
                              </button>
                              <button
                                class="cxd-Button cxd-Button--link cxd-Button--size-xs"
                                type="button"
                              >
                                添加条件组
                              </button>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <span
                    class="cxd-TplField cxd-Form-description"
                  >
                    <span>
                      适合让用户自己拼查询条件，然后后端根据数据生成 query where
                    </span>
                  </span>
                </div>
              </form>
            </div>
            <div
              class="cxd-Panel-footerWrap"
            >
              <div
                class="cxd-Panel-btnToolbar cxd-Panel-footer"
              >
                <button
                  class="cxd-Button cxd-Button--primary cxd-Button--size-default"
                  type="submit"
                >
                  <span>
                    Submit
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
